Add optimizer_for_mobile doc into python api root doc

xcheng16 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41211 Add optimizer_for_mobile doc into python api root doc**

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22543608](https://our.internmc.facebook.com/intern/diff/D22543608)

xcheng16 committed:
Add optimizer_for_mobile doc into python api root doc
Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jlin27 commented on docs/source/mobile_optimizer.rst:
```diff
@@ -0,0 +1,26 @@
+
+
+.. note::
+By default, if optimization blacklist is None or empty, “optimize_for_mobile” will run the following optimizations:
```
Suggestion: Use either italics or code syntax for optimize_for_mobile of quotation marks.

jlin27 commented on docs/source/mobile_optimizer.rst:
```diff
@@ -0,0 +1,26 @@
+4. MobileOptimizerType::REMOVE_DROPOUT: This optimization pass removes dropout and dropout_ nodes from this module when training is false.
+
+
+“optimize_for_mobile” will also invoke freeze_module pass which only preserve only forward method. If you have other method to that needed to be preserved,  add them into the preserved method list and pass into the method.
```
Suggestion: Use either italics or code syntax for optimize_for_mobile of quotation marks.

jlin27 commented on docs/source/mobile_optimizer.rst:
```diff
@@ -0,0 +1,26 @@
+.. note::
+By default, if optimization blacklist is None or empty, “optimize_for_mobile” will run the following optimizations:
+
+1. MobileOptimizerType::CONV_BN_FUSION:  This optimization pass folds Conv2d-BatchNorm2d into Conv2d in forward method of this module and all its submodules. The weight and bias of the Conv2d are correspondingly updated.
```
Suggestion: Link  or bold so it is easier to see. For example:
```
**MobileOptimizerType::CONV_BN_FUSION:**
```

jlin27 commented on docs/source/mobile_optimizer.rst:
```diff
@@ -0,0 +1,26 @@
+
+“optimize_for_mobile” will also invoke freeze_module pass which only preserve only forward method. If you have other method to that needed to be preserved,  add them into the preserved method list and pass into the method.
+
+.. currentmodule:: torch.utils.mobile_optimizer
```
This function isn't rendering for me. Are you seeing it when you build locally?

jlin27 commented:
Hi @xcheng16 - I just have a few suggestions for formatting and code rendering. Here's how it looks when I build locally:  ![screencapture-file-Users-jplin-github-pytorch-docs-build-html-mobile-optimizer-html-2020-07-09-16_57_46](https://user-images.githubusercontent.com/8042156/87105043-ab731c80-c20e-11ea-92d2-f2153520f00c.png)

xcheng16 commented on docs/source/mobile_optimizer.rst:
```diff
@@ -0,0 +1,26 @@
+torch.utils.mobile_optimizer
```
sure, I have this in the local repo, will update it.

xcheng16 committed:
Update on "Add optimizer_for_mobile doc into python api root doc"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

xcheng16 committed:
Update on "Add optimizer_for_mobile doc into python api root doc"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

xcheng16 committed:
Update on "Add optimizer_for_mobile doc into python api root doc"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

xcheng16 commented:
@jlin27 Below is latest version on how it looks like locally.
![Screen Shot 2020-07-10 at 12 34 44 PM](https://user-images.githubusercontent.com/55993203/87191794-0148d280-c2aa-11ea-991a-8e2661144349.png)

kimishpatel commented:
> @jlin27 Below is latest version on how it looks like locally.
> ![Screen Shot 2020-07-10 at 12 34 44 PM](https://user-images.githubusercontent.com/55993203/87191794-0148d280-c2aa-11ea-991a-8e2661144349.png)

Maybe instead of naming pass with MobileOptimizerType::CONV_BN_FUSION, we should just put in bold the name of the pass like how ReLU/Hardtanh fusion is. Otherwise there is a discrepancy between how ReLU/Hardtanh looks vs rest. I guess we have not added an option to disable relu fusion.

It may look like:
```
- **Conv2D + BatchNorm fusion** (blacklisting option `MobileOptimizerType::CONV_BN_FUSION`:  This optimization pass folds ``Conv2d-BatchNorm2d`` into ``Conv2d`` in ``forward`` method of this module and all its submodules. The weight and bias of the ``Conv2d`` are correspondingly updated.
```
Not sure if this is better. But just a suggestion.

xcheng16 commented:
@kimishpatel I like this idea.

xcheng16 committed:
Update on "Add optimizer_for_mobile doc into python api root doc"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

xcheng16 committed:
Update on "Add optimizer_for_mobile doc into python api root doc"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22543608](https://our.internmc.facebook.com/intern/diff/D22543608)

[ghstack-poisoned]
